File: php/src/Resources/CopilotResource.php
<?php

declare(strict_types=1);

namespace JewelMusic\Resources;

/**
 * Copilot resource for AI-powered music generation and assistance
 */
class CopilotResource extends BaseResource
{
    /**
     * Generate AI-powered melody
     *
     * @param array $options Melody generation options
     *                      - style: Musical style (pop, jazz, classical, etc.)
     *                      - genre: Musical genre
     *                      - mood: Desired mood (upbeat, melancholic, energetic, etc.)
     *                      - tempo: Tempo in BPM
     *                      - key: Musical key (C major, A minor, etc.)
     *                      - mode: Musical mode (major, minor, dorian, etc.)
     *                      - timeSignature: Time signature (4/4, 3/4, etc.)
     *                      - duration: Duration in bars
     *                      - instruments: Array of instruments
     *                      - complexity: Complexity level (simple, medium, complex)
     *                      - energy: Energy level (low, medium, high)
     *                      - creativity: Creativity factor (0.0 to 1.0)
     *                      - seed: Optional seed for reproducibility
     * @return array Generated melody data
     */
    public function generateMelody(array $options = []): array
    {
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/melody', $data);
        return $this->extractData($response);
    }

    /**
     * Generate harmony for existing melody
     *
     * @param array $options Harmony generation options
     *                      - melodyId: ID of the melody to harmonize
     *                      - style: Harmony style
     *                      - complexity: Complexity level
     *                      - voicing: Voicing type (close, open, etc.)
     *                      - instruments: Array of instruments
     *                      - creativity: Creativity factor (0.0 to 1.0)
     * @return array Generated harmony data
     */
    public function generateHarmony(array $options = []): array
    {
        if (isset($options['melodyId'])) {
            $this->validateRequired($options, ['melodyId']);
        }
        
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/harmony', $data);
        return $this->extractData($response);
    }

    /**
     * Generate AI-powered lyrics
     *
     * @param array $options Lyrics generation options
     *                      - theme: Lyrical theme
     *                      - genre: Musical genre
     *                      - language: Language code (en, es, fr, etc.)
     *                      - mood: Emotional mood
     *                      - structure: Song structure
     *                      - rhymeScheme: Rhyme scheme (ABAB, AABB, etc.)
     *                      - syllableCount: Syllable pattern
     *                      - inspirationText: Optional inspiration text
     *                      - verseCount: Number of verses
     *                      - chorusCount: Number of choruses
     *                      - bridgeCount: Number of bridges
     *                      - wordsPerLine: Words per line
     *                      - explicitContent: Allow explicit content
     *                      - keywords: Array of keywords to include
     *                      - referenceArtists: Array of reference artists
     * @return array Generated lyrics data
     */
    public function generateLyrics(array $options = []): array
    {
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/lyrics', $data);
        return $this->extractData($response);
    }

    /**
     * Generate a complete song with AI
     *
     * @param array $options Song generation options
     *                      - prompt: Text prompt describing the song
     *                      - melodyId: Optional existing melody ID
     *                      - harmonyId: Optional existing harmony ID
     *                      - lyricsId: Optional existing lyrics ID
     *                      - templateId: Optional template ID
     *                      - style: Musical style
     *                      - duration: Duration in seconds
     *                      - includeVocals: Include vocal generation
     *                      - vocalStyle: Vocal style (male, female, etc.)
     *                      - mixingStyle: Mixing style preset
     *                      - masteringPreset: Mastering preset
     *                      - completionType: Type of completion (intro, outro, bridge, full)
     *                      - addIntro: Add intro section
     *                      - addOutro: Add outro section
     *                      - addBridge: Add bridge section
     * @return array Generated song data
     */
    public function completeSong(array $options = []): array
    {
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/complete-song', $data);
        return $this->extractData($response);
    }

    /**
     * Apply style transfer to existing content
     *
     * @param array $options Style transfer options
     *                      - sourceId: Source content ID (required)
     *                      - targetStyle: Target style to apply (required)
     *                      - intensity: Transfer intensity (0.0 to 1.0)
     *                      - preserveStructure: Preserve original structure
     *                      - preserveTiming: Preserve original timing
     * @return array Style-transferred content
     */
    public function styleTransfer(array $options): array
    {
        $this->validateRequired($options, ['sourceId', 'targetStyle']);
        
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/style-transfer', $data);
        return $this->extractData($response);
    }

    /**
     * Generate chord progression
     *
     * @param array $options Chord progression options
     *                      - key: Musical key (default: C major)
     *                      - style: Musical style (default: pop)
     *                      - complexity: Complexity level (0.0 to 1.0)
     *                      - length: Number of chords
     * @return array Generated chord progression
     */
    public function generateChordProgression(array $options = []): array
    {
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/chord-progression', $data);
        return $this->extractData($response);
    }

    /**
     * Get arrangement suggestions for a track
     *
     * @param array $options Arrangement options
     *                      - trackId: Track ID to analyze (required)
     *                      - targetStyle: Target arrangement style
     * @return array Arrangement suggestions
     */
    public function suggestArrangement(array $options): array
    {
        $this->validateRequired($options, ['trackId']);
        
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/suggest-arrangement', $data);
        return $this->extractData($response);
    }

    /**
     * Analyze genre and suggest improvements
     *
     * @param array $options Genre analysis options
     *                      - trackId: Track ID to analyze (required)
     *                      - targetGenre: Target genre for optimization
     * @return array Genre analysis and suggestions
     */
    public function analyzeGenre(array $options): array
    {
        $this->validateRequired($options, ['trackId']);
        
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/genre-analysis', $data);
        return $this->extractData($response);
    }

    /**
     * Match music to target mood/emotion
     *
     * @param array $options Mood matching options
     *                      - trackId: Track ID to analyze (required)
     *                      - targetMood: Target mood to match (required)
     *                      - adjustmentType: Type of adjustment (tempo, key, arrangement, all)
     * @return array Mood matching suggestions
     */
    public function matchMood(array $options): array
    {
        $this->validateRequired($options, ['trackId', 'targetMood']);
        
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/mood-matching', $data);
        return $this->extractData($response);
    }

    /**
     * Generate variations of existing composition
     *
     * @param array $options Variation options
     *                      - sourceId: Source composition ID (required)
     *                      - variationType: Type of variation (melodic, rhythmic, harmonic, structural)
     *                      - count: Number of variations to generate (1-10)
     * @return array Generated variations
     */
    public function generateVariations(array $options): array
    {
        $this->validateRequired($options, ['sourceId']);
        
        if (isset($options['count'])) {
            if ($options['count'] < 1 || $options['count'] > 10) {
                throw new \InvalidArgumentException('Count must be between 1 and 10');
            }
        }
        
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/generate-variations', $data);
        return $this->extractData($response);
    }

    /**
     * Get AI-powered mastering suggestions
     *
     * @param array $options Mastering options
     *                      - trackId: Track ID to analyze (required)
     *                      - targetPlatform: Target platform for optimization
     * @return array Mastering suggestions
     */
    public function getMasteringSuggestions(array $options): array
    {
        $this->validateRequired($options, ['trackId']);
        
        $data = $this->filterNullValues($options);
        $response = $this->httpClient->post('/copilot/mastering-suggestions', $data);
        return $this->extractData($response);
    }

    /**
     * Get available song templates
     *
     * @param array $filters Template filters
     *                      - genre: Filter by genre
     *                      - mood: Filter by mood
     *                      - duration: Filter by duration
     *                      - style: Filter by style
     * @return array Available templates
     */
    public function getTemplates(array $filters = []): array
    {
        $params = $this->buildParams([], $filters);
        $response = $this->httpClient->get('/copilot/templates', $params);
        return $this->extractData($response);
    }

    /**
     * Get a specific generation by ID
     *
     * @param string $generationId Generation ID
     * @return array Generation data
     */
    public function getGeneration(string $generationId): array
    {
        $response = $this->httpClient->get("/copilot/generations/{$generationId}");
        return $this->extractData($response);
    }

    /**
     * List user's generations with pagination
     *
     * @param int $page Page number
     * @param int $perPage Items per page
     * @param string|null $type Filter by generation type
     * @return array Generations list
     */
    public function listGenerations(int $page = 1, int $perPage = 20, ?string $type = null): array
    {
        $params = [
            'page' => (string)$page,
            'perPage' => (string)$perPage
        ];
        
        if ($type !== null) {
            $params['type'] = $type;
        }
        
        $response = $this->httpClient->get('/copilot/generations', $params);
        return $this->extractData($response);
    }
}

File: ruby/lib/jewelmusic/resources/copilot_resource.rb
# frozen_string_literal: true

module JewelMusic
  module Resources
    ##
    # Copilot resource for AI-powered music generation and assistance
    #
    class CopilotResource < BaseResource
      ##
      # Generate AI-powered melody
      #
      # @param options [Hash] Melody generation options
      # @option options [String] :style Musical style (pop, jazz, classical, etc.)
      # @option options [String] :genre Musical genre
      # @option options [String] :mood Desired mood (upbeat, melancholic, energetic, etc.)
      # @option options [Integer] :tempo Tempo in BPM
      # @option options [String] :key Musical key (C major, A minor, etc.)
      # @option options [String] :mode Musical mode (major, minor, dorian, etc.)
      # @option options [String] :time_signature Time signature (4/4, 3/4, etc.)
      # @option options [Integer] :duration Duration in bars
      # @option options [Array<String>] :instruments Array of instruments
      # @option options [String] :complexity Complexity level (simple, medium, complex)
      # @option options [String] :energy Energy level (low, medium, high)
      # @option options [Float] :creativity Creativity factor (0.0 to 1.0)
      # @option options [String] :seed Optional seed for reproducibility
      # @return [Hash] Generated melody data
      #
      def generate_melody(options = {})
        data = filter_nil_values(options)
        response = http_client.post('/copilot/melody', data)
        extract_data(response)
      end

      ##
      # Generate harmony for existing melody
      #
      # @param options [Hash] Harmony generation options
      # @option options [String] :melody_id ID of the melody to harmonize
      # @option options [String] :style Harmony style
      # @option options [String] :complexity Complexity level
      # @option options [String] :voicing Voicing type (close, open, etc.)
      # @option options [Array<String>] :instruments Array of instruments
      # @option options [Float] :creativity Creativity factor (0.0 to 1.0)
      # @return [Hash] Generated harmony data
      #
      def generate_harmony(options = {})
        validate_required(options, [:melody_id]) if options[:melody_id]
        
        data = filter_nil_values(options)
        response = http_client.post('/copilot/harmony', data)
        extract_data(response)
      end

      ##
      # Generate AI-powered lyrics
      #
      # @param options [Hash] Lyrics generation options
      # @option options [String] :theme Lyrical theme
      # @option options [String] :genre Musical genre
      # @option options [String] :language Language code (en, es, fr, etc.)
      # @option options [String] :mood Emotional mood
      # @option options [String] :structure Song structure
      # @option options [String] :rhyme_scheme Rhyme scheme (ABAB, AABB, etc.)
      # @option options [String] :syllable_count Syllable pattern
      # @option options [String] :inspiration_text Optional inspiration text
      # @option options [Integer] :verse_count Number of verses
      # @option options [Integer] :chorus_count Number of choruses
      # @option options [Integer] :bridge_count Number of bridges
      # @option options [Integer] :words_per_line Words per line
      # @option options [Boolean] :explicit_content Allow explicit content
      # @option options [Array<String>] :keywords Keywords to include
      # @option options [Array<String>] :reference_artists Reference artists
      # @return [Hash] Generated lyrics data
      #
      def generate_lyrics(options = {})
        data = filter_nil_values(options)
        response = http_client.post('/copilot/lyrics', data)
        extract_data(response)
      end

      ##
      # Generate a complete song with AI
      #
      # @param options [Hash] Song generation options
      # @option options [String] :prompt Text prompt describing the song
      # @option options [String] :melody_id Optional existing melody ID
      # @option options [String] :harmony_id Optional existing harmony ID
      # @option options [String] :lyrics_id Optional existing lyrics ID
      # @option options [String] :template_id Optional template ID
      # @option options [String] :style Musical style
      # @option options [Integer] :duration Duration in seconds
      # @option options [Boolean] :include_vocals Include vocal generation
      # @option options [String] :vocal_style Vocal style (male, female, etc.)
      # @option options [String] :mixing_style Mixing style preset
      # @option options [String] :mastering_preset Mastering preset
      # @option options [String] :completion_type Type of completion (intro, outro, bridge, full)
      # @option options [Boolean] :add_intro Add intro section
      # @option options [Boolean] :add_outro Add outro section
      # @option options [Boolean] :add_bridge Add bridge section
      # @return [Hash] Generated song data
      #
      def complete_song(options = {})
        data = filter_nil_values(options)
        response = http_client.post('/copilot/complete-song', data)
        extract_data(response)
      end

      ##
      # Apply style transfer to existing content
      #
      # @param options [Hash] Style transfer options
      # @option options [String] :source_id Source content ID (required)
      # @option options [String] :target_style Target style to apply (required)
      # @option options [Float] :intensity Transfer intensity (0.0 to 1.0)
      # @option options [Boolean] :preserve_structure Preserve original structure
      # @option options [Boolean] :preserve_timing Preserve original timing
      # @return [Hash] Style-transferred content
      #
      def style_transfer(options)
        validate_required(options, [:source_id, :target_style])
        
        data = filter_nil_values(options)
        response = http_client.post('/copilot/style-transfer', data)
        extract_data(response)
      end

      ##
      # Generate chord progression
      #
      # @param options [Hash] Chord progression options
      # @option options [String] :key Musical key (default: C major)
      # @option options [String] :style Musical style (default: pop)
      # @option options [Float] :complexity Complexity level (0.0 to 1.0)
      # @option options [Integer] :length Number of chords
      # @return [Hash] Generated chord progression
      #
      def generate_chord_progression(options = {})
        data = filter_nil_values(options)
        response = http_client.post('/copilot/chord-progression', data)
        extract_data(response)
      end

      ##
      # Get arrangement suggestions for a track
      #
      # @param options [Hash] Arrangement options
      # @option options [String] :track_id Track ID to analyze (required)
      # @option options [String] :target_style Target arrangement style
      # @return [Hash] Arrangement suggestions
      #
      def suggest_arrangement(options)
        validate_required(options, [:track_id])
        
        data = filter_nil_values(options)
        response = http_client.post('/copilot/suggest-arrangement', data)
        extract_data(response)
      end

      ##
      # Analyze genre and suggest improvements
      #
      # @param options [Hash] Genre analysis options
      # @option options [String] :track_id Track ID to analyze (required)
      # @option options [String] :target_genre Target genre for optimization
      # @return [Hash] Genre analysis and suggestions
      #
      def analyze_genre(options)
        validate_required(options, [:track_id])
        
        data = filter_nil_values(options)
        response = http_client.post('/copilot/genre-analysis', data)
        extract_data(response)
      end

      ##
      # Match music to target mood/emotion
      #
      # @param options [Hash] Mood matching options
      # @option options [String] :track_id Track ID to analyze (required)
      # @option options [String] :target_mood Target mood to match (required)
      # @option options [String] :adjustment_type Type of adjustment (tempo, key, arrangement, all)
      # @return [Hash] Mood matching suggestions
      #
      def match_mood(options)
        validate_required(options, [:track_id, :target_mood])
        
        data = filter_nil_values(options)
        response = http_client.post('/copilot/mood-matching', data)
        extract_data(response)
      end

      ##
      # Generate variations of existing composition
      #
      # @param options [Hash] Variation options
      # @option options [String] :source_id Source composition ID (required)
      # @option options [String] :variation_type Type of variation (melodic, rhythmic, harmonic, structural)
      # @option options [Integer] :count Number of variations to generate (1-10)
      # @return [Hash] Generated variations
      #
      def generate_variations(options)
        validate_required(options, [:source_id])
        
        if options[:count] && (options[:count] < 1 || options[:count] > 10)
          raise ArgumentError, 'Count must be between 1 and 10'
        end
        
        data = filter_nil_values(options)
        response = http_client.post('/copilot/generate-variations', data)
        extract_data(response)
      end

      ##
      # Get AI-powered mastering suggestions
      #
      # @param options [Hash] Mastering options
      # @option options [String] :track_id Track ID to analyze (required)
      # @option options [String] :target_platform Target platform for optimization
      # @return [Hash] Mastering suggestions
      #
      def get_mastering_suggestions(options)
        validate_required(options, [:track_id])
        
        data = filter_nil_values(options)
        response = http_client.post('/copilot/mastering-suggestions', data)
        extract_data(response)
      end

      ##
      # Get available song templates
      #
      # @param filters [Hash] Template filters
      # @option filters [String] :genre Filter by genre
      # @option filters [String] :mood Filter by mood
      # @option filters [String] :duration Filter by duration
      # @option filters [String] :style Filter by style
      # @return [Hash] Available templates
      #
      def get_templates(filters = {})
        params = build_params({}, filters)
        response = http_client.get('/copilot/templates', params)
        extract_data(response)
      end

      ##
      # Get a specific generation by ID
      #
      # @param generation_id [String] Generation ID
      # @return [Hash] Generation data
      #
      def get_generation(generation_id)
        validate_required({ generation_id: generation_id }, [:generation_id])
        response = http_client.get("/copilot/generations/#{generation_id}")
        extract_data(response)
      end

      ##
      # List user's generations with pagination
      #
      # @param page [Integer] Page number
      # @param per_page [Integer] Items per page
      # @param type [String, nil] Filter by generation type
      # @return [Hash] Generations list
      #
      def list_generations(page = 1, per_page = 20, type = nil)
        params = {
          page: page.to_s,
          per_page: per_page.to_s
        }
        
        params[:type] = type if type
        
        response = http_client.get('/copilot/generations', params)
        extract_data(response)
      end
    end
  end
end